nutils/_util.py

```diff
@@ -1082,4 +1082,39 @@ def shallow_replace(func, *funcargs, **funckwargs):
     return rstack[0]
 
 
+def tree_walk(visit_node, /, *roots):
+    '''calls ``visit_node`` for every node in ``roots``
+
+    The callable ``visit_node`` takes one node as argument and should return
+    the child nodes to visit or ``None``. Every identical node is visited only
+    once.
+
+    Examples
+    --------
+
+    >>> def visit(node):
+    ...     if isinstance(node, list):
+    ...         return node
+    ...     else:
+    ...         print(node)
+    >>> tree_walk(visit, [1, [2, 3]])
+    3
+    2
+    1
+    '''
+
+    queue = []
+    seen = set()
+    for root in roots:
+        if id(root) not in seen:
+            queue.append(root)
+            seen.add(id(root))
+    while queue:
+        node = queue.pop()
+        for dep in visit_node(node) or ():
+            if id(dep) not in seen:
+                queue.append(dep)
+                seen.add(id(dep))
```

Comment:
Doesn't this have a simpler implementation in
```python
stack = IDSet(roots)
while stack:
    stack.update(visit_node(stack.pop()) or ())
```
This gets even cleaner if the visitor returns an empty list to terminate instead of `None`, which also opens it up to generators.
(Note I renamed queue to stack since we're appending and popping from the same side)

Reply:
Every node should be visited only once. That's where `seen` is for.

Reply:
Right, didn't think of the items that are popped and forgotten.